### ISSUE TITLE
XSDSchema11 constraints validation failure

### ISSUE THREAD

**Issue by derreter** [sissaschool/xmlschema#160]
Hi,

The program is not validating constraint rules, according to the prints attached. Prints are just a few rules, but they should all fail.
Note: I am testing with the latest version (1.0.18)

Thanks again.

[testFiles.zip](https://github.com/sissaschool/xmlschema/files/4026035/testFiles.zip)

**Comment by brunato** [comment]
Hi,
as I see the sample file (saft_idemo599999999.xml) has 4 wrong keys with an underscore at left: "_1111", "_FAT01", "FAT01", "_SB18" and "_CA3". Removing the underscore from these keys the file is valid.

If don't know how the errors (the reasons ...) that you have attached were obtained, but my test with the v1.0.18 extracts all the errors:

```python
>>> import xmlschema
>>> xs = xmlschema.XMLSchema11("saftpt1.04_01.xsd")
>>> errors = list(xs.iter_errors("saft_idemo599999999.xml"))
>>> len(errors)
63
>>> for e in errors:
...   print(e.reason)
... 
Key 'ns:GeneralLedgerEntriesDebitLineAccountIDConstraint' with value ('1111',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesDebitLineAccountIDConstraint' with value ('1111',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesDebitLineAccountIDConstraint' with value ('1111',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesDebitLineAccountIDConstraint' with value ('1111',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:GeneralLedgerEntriesSupplierIDConstraint' with value ('SB18',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceCustomerIDConstraint' with value ('CA3',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:InvoiceProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementSupplierIDConstraint' with value ('SB18',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:StockMovementProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
Key 'ns:WorkDocumentDocumentProductCodeConstraint' with value ('FAT01',) not found for identity constraint of element 'ns:AuditFile'.
```

**Comment by derreter** [comment]
Hi,

The existing errors were caused by me to test.
I was using this code: 

>>> import xmlschema
>>> xs = xmlschema.XMLSchema11("saftpt1.04_01.xsd")
>>> errors = [e for e in xmlschema.iter_errors("saft_idemo599999999.xml", schema=xs, lazy=True)]
>>> len(errors)
0

Based on a response from you to another user, but as it returns zero errors
It was based on what I reported as not validating the constraints.

Thanks

**Comment by derreter** [comment]
Hi,
It's me again ...

After a few more tests, it concludes that if the file is a bit larger, the validation does not end.
I send the test files. The xml only has one error shown in the image.

Code used:  errors = list(xs.iter_errors("xpto.xml"))

Thanks.

[testFiles.zip](https://github.com/sissaschool/xmlschema/files/4032381/testFiles.zip)

**Comment by brunato** [comment]
> 
> ```python
> >>> import xmlschema
> >>> xs = xmlschema.XMLSchema11("saftpt1.04_01.xsd")
> >>> errors = [e for e in xmlschema.iter_errors("saft_idemo599999999.xml", schema=xs, lazy=True)]
> >>> len(errors)
> 0
> ```
Ok, more clear now...

The problem is about limitation of identity constraints in *lazy* mode, because standard XPath operations cannot work with *lazy* (partial) ElementTree structures. Currently there is no mode for checking xs:keyrefs. Maybe this part need to be rewritten using XPath on schemas to labeling target XsdElements (the elements selected by an identity constraint) and then pass a key_map for collecting values during validation (and probably other things). Not so simple ...

**Comment by brunato** [comment]
> After a few more tests, it concludes that if the file is a bit larger, the validation does not end.

The validation ends for me (2 errors found) but is very slow (about half an hour). It stucks at identity constraints validation (that are validated at the end of the root element ... so at the end of the validation).

```
Xsd11Unique(name='ns:AccountIDConstraint')
Xsd11Keyref(name='ns:GroupingCodeConstraint', refer='ns:AccountIDConstraint')
Xsd11Unique(name='ns:CustomerIDConstraint')
Xsd11Unique(name='ns:SupplierIDConstraint')
Xsd11Unique(name='ns:ProductCodeConstraint')
Xsd11Keyref(name='ns:GeneralLedgerEntriesDebitLineAccountIDConstraint', refer='ns:AccountIDConstraint')
Xsd11Keyref(name='ns:GeneralLedgerEntriesCreditLineAccountIDConstraint', refer='ns:AccountIDConstraint')
Xsd11Keyref(name='ns:GeneralLedgerEntriesCustomerIDConstraint', refer='ns:CustomerIDConstraint')
Xsd11Unique(name='ns:GeneralLedgerEntriesJournalIdConstraint')
Xsd11Keyref(name='ns:GeneralLedgerEntriesSupplierIDConstraint', refer='ns:SupplierIDConstraint')
Xsd11Unique(name='ns:GeneralLedgerEntriesTransactionIdConstraint')
Xsd11Unique(name='ns:InvoiceNoConstraint')
Xsd11Keyref(name='ns:InvoiceCustomerIDConstraint', refer='ns:CustomerIDConstraint')
Xsd11Keyref(name='ns:InvoiceProductCodeConstraint', refer='ns:ProductCodeConstraint')
Xsd11Unique(name='ns:DocumentNumberConstraint')
Xsd11Keyref(name='ns:StockMovementCustomerIDConstraint', refer='ns:CustomerIDConstraint')
Xsd11Keyref(name='ns:StockMovementSupplierIDConstraint', refer='ns:SupplierIDConstraint')
Xsd11Keyref(name='ns:StockMovementProductCodeConstraint', refer='ns:ProductCodeConstraint')
Xsd11Unique(name='ns:WorkDocumentDocumentNumberConstraint')
Xsd11Keyref(name='ns:WorkDocumentDocumentCustomerIDConstraint', refer='ns:CustomerIDConstraint')
Xsd11Keyref(name='ns:WorkDocumentDocumentProductCodeConstraint', refer='ns:ProductCodeConstraint')
Xsd11Unique(name='ns:PaymentPaymentRefNoConstraint')
Xsd11Keyref(name='ns:PaymentPaymentRefNoCustomerIDConstraint', refer='ns:CustomerIDConstraint')
```
These XPath selectors, that spread on almost the whole tree, are very slow. So this confirms the need to rewrite identity constraints checks ...

**Comment by brunato** [comment]
Rewrite of identity checks almost done. Not so difficult to do as I tought. It will be based on *collections.Counters* as ID/IDREF to save memory. 
Seems to be also pretty fast (about 10 seconds on my laptop for your big file).
Will be available with release v1.1.0.

**Comment by derreter** [comment]
Ok,
Thanks.

**Comment by brunato** [comment]
This is a test with speed on both XML files, with normal and with lazy validation:

```python
>>> import xmlschema
>>> from datetime import datetime as dt
>>> xs = xmlschema.XMLSchema11("saftpt1.04_01.xsd")
>>> st = dt.now(); len(list(xmlschema.iter_errors("saft_idemo599999999.xml", schema=xs))), dt.now() - st
(63, datetime.timedelta(microseconds=585027))
>>> st = dt.now(); len(list(xmlschema.iter_errors("saft_idemo599999999.xml", schema=xs, lazy=True))), dt.now() - st
(63, datetime.timedelta(microseconds=548806))
>>> st = dt.now(); len(list(xmlschema.iter_errors("XPTO.XML", schema=xs))), dt.now() - st
(2, datetime.timedelta(seconds=9, microseconds=432212))
>>> st = dt.now(); len(list(xmlschema.iter_errors("XPTO.XML", schema=xs, lazy=True))), dt.now() - st
(2, datetime.timedelta(seconds=9, microseconds=349851))
```
The 2nd file has these two errors:
```python
>>> for e in xmlschema.iter_errors("XPTO.XML", schema=xs): e.reason
... 
"Referred key 'ns:CustomerIDConstraint' with value ('01095',) not found"
"Referred key 'ns:CustomerIDConstraint' with value ('01095',) not found"
```
is it correct?

**Comment by derreter** [comment]
No, its not correct. 
Its the same Costumer referenced in two places from same table (Transaction).
For me, its one error. (same for Altova xmlspy)

Thanks.

**Comment by brunato** [comment]
> No, its not correct.
> Its the same Costumer referenced in two places from same table (Transaction).
> For me, its one error. (same for Altova xmlspy)
> 
Ok, reworked to include the repetition into error's reason:

```python
>>> import xmlschema
>>> xs = xmlschema.XMLSchema11("saftpt1.04_01.xsd")
>>> for e in xmlschema.iter_errors("XPTO.XML", schema=xs): e.reason
... 
"Value ('01095',) not found for key ns:CustomerIDConstraint (2 times)"
```

**Comment by brunato** [comment]
Hi,
this should be fixed with release v1.1.0 (requires Py35+).
Bye
